### DISCUSSION TITLE
RESET reason found out

### DISCUSSION THREAD

**Discussion by kendo55** [earlephilhower/arduino-pico#1825] [Q&A]
on esp8266 we had the opportunity to ask for the RESET reason after reboot.
is there an equivalent on PICO?

on esp8266
--------------
  rst_info *resetInfo;
  resetInfo = ESP.getResetInfoPtr();
  uint16_t reset_reason = resetInfo->reason;

  "0 Power reboot\n"
  "1 Hardware WDT reset\n"
  "2 Fatal exception\n"
  "3 Software watchdog reset\n"
  "4 Software reset\n"
  "5 Deep-sleep\n"
  "6 Hardware reset\n";

**Comment by earlephilhower** [comment]
Because things like WDT and reset are all HW driven and not really Arduino core related, I think you will have better luck on the [RPI forums](https://forums.raspberrypi.com/viewforum.php?f=143&sid=89f1ff8bfafb653acc8dd2715624ff48) or [Pico-SDK](https://github.com/raspberrypi/pico-sdk) repo.

**Comment by kendo55** [comment]
my proposal for a additional function in the RP2040 Helper Class
------------------------------------------------------------------------
 void rp2040.resetReason()


````

#include "hardware/regs/vreg_and_chip_reset.h"
  // C:\Pico\PicoSDK_v1.5.0\pico-sdk\src\rp2040\hardware_regs\include\hardware\regs\vreg_and_chip_reset.h
#include "hardware/regs/watchdog.h"

const char reset_reason_txt[]  = 
"Reset Reason:\n"
" 0 unknown\n"
" 1 power-on or brown-out\n"
" 2 Hardware PIN Reset\n"
" 3 Software reset over WD\n"
" 4 Hardware watchdog timer reset\n";

uint8_t resetReason(bool talk) {
//-------------------- 
  io_rw_32 *reset_reason_reg = (io_rw_32 *) (VREG_AND_CHIP_RESET_BASE + VREG_AND_CHIP_RESET_CHIP_RESET_OFFSET);
  uint8_t nResReason = 0;
  
  uint8_t nWD_Reason = WD_Reason();  // SDK func watchdog_caused_reboot() does not deliver the reason
  
  if (nWD_Reason > 0){
    nResReason = nWD_Reason +  2;  // 3 and 4
  }else{   

    if ((*reset_reason_reg & VREG_AND_CHIP_RESET_CHIP_RESET_HAD_POR_BITS)) { 
      // POR: Last reset was from the power-on reset or brown-out detection
      nResReason = 1;
    }
    if ((*reset_reason_reg & VREG_AND_CHIP_RESET_CHIP_RESET_HAD_RUN_BITS)) { 
      // RUN: Last reset was from the RUN pin
      nResReason = 2;
    }  
  }
  if (talk){
    Serial1.printf("RESET Reason = %d\n", nResReason);
    Serial1.print(reset_reason_txt);
  }
  return nResReason;
}

uint8_t WD_Reason(void) {
//-----------------------
// Watchdog reason
  io_rw_32 *WD_reason_reg = (io_rw_32 *) (WATCHDOG_BASE + WATCHDOG_REASON_OFFSET);
  uint8_t nWD_Reason = 0;
  
  if ((*WD_reason_reg & WATCHDOG_REASON_FORCE_BITS)) { 
    // WATCHDOG_REASON_FORCE
    nWD_Reason = 1;
  } 
  if ((*WD_reason_reg & WATCHDOG_REASON_TIMER_BITS)) { 
    // WATCHDOG_REASON_TIMER
    nWD_Reason = 2;
  }
  return nWD_Reason;
}


````
output:
----------------------------
RESET Reason = 3
Reset Reason:
 0 unknown
 1 power-on or brown-out
 2 Hardware PIN Reset
 3 Software reset over WD
 4 Hardware watchdog timer reset

**Reply by earlephilhower** [reply]
I think we need to drop the printouts, make that return an `enum` and not just a bare number, but it seems like a good idea for a PR! :+1:

**Reply by kendo55** [reply]
Of course you are right.
I'm an old engineer who started to programing Microprocessors with the Intel 4040.
And it's now the last chance to come friendly with enums :-)

**Comment by palmerr23** [comment]
Kendo55,

Are you going to follow through with this addition? 
I had a similar idea and will turn it into a pull request if you aren't going to. See also [https://github.com/earlephilhower/arduino-pico/discussions/1843](url)

Richard
